### PR TITLE
Add support for a 2.2 EarlGrey.swift version in the EarlGrey gem.

### PR DESCRIPTION
**Pull request by tirodkar**
This still needs to be pushed out to the official gem. It worked perfectly fine locally.

Fixes https://github.com/google/EarlGrey/issues/212

### CODE REVIEW

File: gem/lib/earlgrey/cli.rb [hunk 1]
```diff
@@ -19,11 +19,12 @@ class CLI < Thor
     package_name 'EarlGrey'
 
     desc 'install', 'Installs EarlGrey into an Xcode unit test target'
+    method_option :carthage, type: :boolean, default: true
     method_option :project, aliases: '-p', type: :string, required: false, desc: 'Project'
-    method_option :target,  aliases: '-t', type: :string, required: true, desc: 'EarlGreyTestTarget'
     method_option :scheme,  aliases: '-s', type: :string, required: false, desc: 'EarlGreyTestTarget.xcscheme'
     method_option :swift,    type: :boolean, default: true
-    method_option :carthage, type: :boolean, default: true
+    method_option :swift_version, type: :numeric, default: 3.0, :enum => %w(2.2, 3.0)
```
Comment (bootstraponline) [inline gem/lib/earlgrey/cli.rb]: Numerics in Thor are parsed as floats or integers. This doesn't work for point releases of Swift. For example, the upcoming [Swift `3.0.1`](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20160725/025676.html).

The correct datatype is string.

Comment (bootstraponline) [inline gem/lib/earlgrey/cli.rb]:
```ruby
%w(2.2, 3.0)
```

The above creates a corrupted first version value. Observe:

```
> %w(2.2, 3.0)
=> ["2.2,", "3.0"]
```

There shouldn't be a comma in element 0. The space delimiter is what's used, not a comma.

File: gem/lib/earlgrey/cli.rb [hunk 2]
```diff
@@ -36,7 +37,7 @@ def install
 
       # CLI will never use Cocoapod's `post_install do |installer|`
       podfile_installer = nil
-      opts = { swift: o[SWIFT], carthage: o[CARTHAGE] }
+      opts = { swift: o[SWIFT], carthage: o[CARTHAGE], swift_version: o[SWIFT_VERSION] }
```
Comment (bootstraponline) [inline gem/lib/earlgrey/cli.rb]: I don't see `SWIFT_VERSION` defined anywhere in this pull request?

Comment (bootstraponline) [inline gem/lib/earlgrey/cli.rb]: Also this is why the Ruby tests are failing:

>  uninitialized constant EarlGrey::CLI::SWIFT_VERSION
>
> https://travis-ci.org/google/EarlGrey/jobs/150811297

File: gem/lib/earlgrey/configure_earlgrey.rb [hunk 3]
```diff
@@ -66,6 +66,7 @@ def puts_yellow(string)
     def set_defaults(installer, project_name, test_target_name, scheme_file, opts = {})
       @swift = opts.fetch(:swift, false)
       @carthage = opts.fetch(:carthage, false)
+      @swift_version = opts.fetch(:swift_version, false)
```
Comment (bootstraponline) [inline gem/lib/earlgrey/configure_earlgrey.rb]: False is not a valid default value for `:swift_version`. `:swift` and `:carthage` are booleans which is why they default to false. `:swift_version` should default to a valid version of Swift.

File: gem/lib/earlgrey/configure_earlgrey.rb [hunk 4]
```diff
@@ -353,7 +354,8 @@ def copy_swift_files
       dst_header = File.join(dst_root, src_header_name)
 
       src_swift_name = 'EarlGrey.swift'
-      src_swift = File.join(src_root, src_swift_name)
+      src_swift = File.join(src_root, "Swift-" + "#{swift_version}/" + src_swift_name)
```
Comment (bootstraponline) [inline gem/lib/earlgrey/configure_earlgrey.rb]: When using interpolation (`#{}`), you don't use +. Also the path separator should never be hard coded as that's the reason we use the Join method. See the corrected code below.

```ruby
 File.join(src_root, "Swift-#{swift_version}", src_swift_name)
```

**Comment by bootstraponline** [comment]
There are no specs. I think at a minimum, we want specs that verify the correct version of EarlGrey.swift is copied over for:

- The default case
- Version 2.2
- Version 3.0

File: gem/lib/earlgrey/cli.rb (hunk 2 above)
Comment (tirodkar) [inline gem/lib/earlgrey/cli.rb]: Done. Thanks!

File: gem/lib/earlgrey/cli.rb [hunk 5]
```diff
@@ -19,24 +19,26 @@ class CLI < Thor
...
 
     desc 'install', 'Installs EarlGrey into an Xcode unit test target'
+    method_option :carthage, type: :boolean, default: true
     method_option :project, aliases: '-p', type: :string, required: false, desc: 'Project'
-    method_option :target,  aliases: '-t', type: :string, required: true, desc: 'EarlGreyTestTarget'
     method_option :scheme,  aliases: '-s', type: :string, required: false, desc: 'EarlGreyTestTarget.xcscheme'
-    method_option :swift,    type: :boolean, default: true
-    method_option :carthage, type: :boolean, default: true
+    method_option :swift, type: :boolean, default: true
+    method_option :swift_version, type: :string, default: '3.0', enum: ['2.2', '3.0']
```
Comment (bootstraponline) [inline gem/lib/earlgrey/cli.rb]: This is a cleaner array syntax ([as mentioned in the Thor wiki](https://github.com/erikhuda/thor/wiki/Method-Options):

```ruby
%w(2.2 3.0)
```

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import EarlGrey
+
+func grey_allOfMatchers(args: AnyObject...) -> GREYMatcher! {
```
Comment (bootstraponline) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: per https://github.com/google/EarlGrey/pull/216 I think these should be public?

File: gem/spec/configure_earlgrey_spec.rb [hunk 7]
```diff
@@ -22,8 +22,22 @@
...
 
-  it 'configures for cocoapods' do
-    diff_project cocoapods_after, %w[install -t AutoEarlGrey --no-carthage]
+  it 'configures for cocoapods with default swift version' do
+    diff_project cocoapods_after_default, %w[install -t AutoEarlGrey --no-carthage]
+    expect(EarlGrey.swift).to eq(true)
+    expect(EarlGrey.carthage).to eq(false)
+  end
+
+  it 'configures for cocoapods with Swift 2.2' do
```
Comment (bootstraponline) [inline gem/spec/configure_earlgrey_spec.rb]: technically this works however we'll potentially end up with 6 fixture projects

2 (cocoapods + carthage) * 3 (swift default, swift 2, swift 3) = 6 

I'd rather keep one baseline for both cocoapods & carthage then dynamically modify the data in memory with the correct path. The only change is the file path so I think this is reasonable.

**Comment by bootstraponline** [comment]
Actually, I'm not seeing any difference between the fixtures at all.... this passes for me:

```diff
diff --git a/gem/spec/configure_earlgrey_spec.rb b/gem/spec/configure_earlgrey_spec.rb
index e4cdf59..b505959 100644
--- a/gem/spec/configure_earlgrey_spec.rb
+++ b/gem/spec/configure_earlgrey_spec.rb
@@ -29,14 +29,14 @@ describe 'configure_earlgrey' do
   end
 
   it 'configures for cocoapods with Swift 2.2' do
-    diff_project cocoapods_after_swift_2,
+    diff_project cocoapods_after_default,
                  %w[install -t AutoEarlGrey --no-carthage --swift_version=2.2]
     expect(EarlGrey.swift).to eq(true)
     expect(EarlGrey.carthage).to eq(false)
   end
 
   it 'configures for cocoapods with Swift 3' do
-    diff_project cocoapods_after_swift_3,
+    diff_project cocoapods_after_default,
                  %w[install -t AutoEarlGrey --no-carthage --swift_version=3.0]
     expect(EarlGrey.swift).to eq(true)
     expect(EarlGrey.carthage).to eq(false)
```

which makes sense because regardless of the source file name, the destination file name is always `EarlGrey.swift` so there's no change.

**Comment by tirodkar** [comment]
Yeah, the file is still EarlGrey.swift so there shouldn't be any change whatsoever there.

**Comment by khandpur** [comment]
You'd need to rebase this.

File: gem/lib/earlgrey/cli.rb (hunk 5 above)
Comment (tirodkar) [inline gem/lib/earlgrey/cli.rb]: That actually threw me an error on `rake`, asking me to move to the hash syntax.

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift (hunk 6 above)
Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Yep! Thanks.

File: gem/lib/earlgrey/cli.rb (hunk 5 above)
Comment (bootstraponline) [inline gem/lib/earlgrey/cli.rb]: Oh, that's just a lint check. Use `%w[2.2 3.0]` instead.

In Ruby the token after the w can be anything.

```
%w*a b*
=> ["a", "b"]
```

**Comment by tirodkar** [comment]
Rebased and updated. I removed the extra projects and instead have an extra parameter passed to the project_diff file.

**Comment by bootstraponline** [comment]
> confirm that they're okay with these commits being contributed to Google.

I'm okay with these commits being contributed to Google & I've signed the CLA. 😄

**Comment by bootstraponline** [comment]
Looks good to merge.

**Comment by khandpur** [comment]
resolving CLA status.

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+public func GREYAssertTrue(@autoclosure expression: () -> BooleanType, reason: String) {
+  GREYAssert(expression().boolValue,
+    reason,
+    details: "Expected the boolean expression to be true")
+}
+
+public func GREYAssertFalse(@autoclosure expression: () -> BooleanType, reason: String) {
+  GREYAssert(!expression().boolValue,
+    reason,
+    details: "Expected the boolean expression to be true")
```
Comment (sid-github) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: s/true/false

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+public func grey_allOfMatchers(args: AnyObject...) -> GREYMatcher! {
+  return GREYAllOf.init(matchers: args)
+}
+
+public func grey_anyOfMatchers(args: AnyObject...) -> GREYMatcher! {
+  return GREYAnyOf.init(matchers: args)
+}
+
+public func EarlGrey() -> EarlGreyImpl! {
+  return EarlGreyImpl.invokedFromFile(__FILE__, lineNumber: __LINE__)
```
Comment (sid-github) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: shouldnt these be ##file  and ##line?
I am comparing with https://github.com/google/EarlGrey/blob/master/gem/lib/earlgrey/files/EarlGrey.swift

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  GREYAssert(expression() != nil, reason, details: "Expected expression to be not nil")
+}
+
+public func GREYAssertNil(@autoclosure expression: () -> Any?, reason: String) {
+  GREYAssert(expression() == nil, reason, details: "Expected expression to be nil")
+}
+
+public func GREYAssertEqual<T : Equatable>(@autoclosure left: () -> T?,
+  @autoclosure _ right: () -> T?, reason: String) {
+    GREYAssert(left() == right(), reason, details: "Expeted left term to be equal to right term")
```
Comment (sid-github) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: *Expected

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+public func GREYFail(reason: String, details: String) {
+  greyFailureHandler.handleException(GREYFrameworkException(name: kGREYAssertionFailedException,
+    reason: reason),
+    details: details)
+}
+
+private func GREYAssert(@autoclosure expression: () -> BooleanType,
+  _ reason: String, details: String) {
+    //GREYSetCurrentAsFailable()
```
Comment (sid-github) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: why is this commented out?

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift (hunk 9 above)
Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Nope this for 2.2

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift (hunk 11 above)
Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Removing.

Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Removed.

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift (hunk 10 above)
Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Done.

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift (hunk 8 above)
Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Done.

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+private func GREYAssert(@autoclosure expression: () -> BooleanType,
+  _ reason: String, details: String) {
+    GREYSetCurrentAsFailable()
+    if !expression().boolValue {
+      greyFailureHandler.handleException(GREYFrameworkException(name: kGREYAssertionFailedException,
+        reason: reason),
+        details: details)
+    }
```
Comment (sid-github) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: indent -2

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  GREYAssert(expression() != nil, reason, details: "Expected expression to be not nil")
+}
+
+public func GREYAssertNil(@autoclosure expression: () -> Any?, reason: String) {
+  GREYAssert(expression() == nil, reason, details: "Expected expression to be nil")
+}
+
+public func GREYAssertEqual<T : Equatable>(@autoclosure left: () -> T?,
+  @autoclosure _ right: () -> T?, reason: String) {
+    GREYAssert(left() == right(), reason, details: "Expected left term to be equal to right term")
```
Comment (sid-github) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: indent -2

Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Done.

File: gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift (hunk 12 above)
Comment (tirodkar) [inline gem/lib/earlgrey/files/Swift-2.2/EarlGrey.swift]: Done.

**Pull request merged**